source/common/router/req_header_formatter.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include <cstdint>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+
+#include "common/common/logger.h"
+#include "common/http/access_log/access_log_formatter.h"
+#include "common/json/json_loader.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * Interface for all types of header formatters used for custom request headers.
+ */
+class HeaderFormatter {
+public:
+  virtual ~HeaderFormatter() {}
+
+  virtual const std::string
+  format(const Envoy::Http::AccessLog::RequestInfo& request_info) const PURE;
+};
+
+typedef std::shared_ptr<HeaderFormatter> HeaderFormatterPtr;
```

Comment:
When i change this to the unique_ptr, i add move semantics in req_header_formatter.cc as follows:

RequestHeaderParser RequestHeaderParser::parse(const Json::Object& config) {
  RequestHeaderParser request_header_parser;
  std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>
      header_formatter_map;
  if (config.hasObject("request_headers_to_add")) {
    std::vector<Json::ObjectSharedPtr> request_headers =
        config.getObjectArray("request_headers_to_add");
    for (const Json::ObjectSharedPtr& header : request_headers) {
      if (!header->getString("key").empty() && !header->getString("value").empty()) {
        ENVOY_LOG(debug, "adding key {} to header formatter map", header->getString("key"));
        HeaderFormatterPtr header_formatter =
            RequestHeaderParser::parseInternal(header->getString("value"));
        request_header_parser.header_formatter_map_.emplace(
            Http::LowerCaseString(header->getString("key")), **std::move(header_formatter)**);
      }
    }
  }

  return request_header_parser;

However, i see the following compilation error
error: call to implicitly-deleted copy constructor of 'std::pair<const Envoy::Http::LowerCaseString, std::unique_ptr<Envoy::Router::HeaderFormatter, std::default_delete<Envoy::Router::HeaderFormatter> > >'